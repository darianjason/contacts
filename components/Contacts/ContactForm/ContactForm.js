// @flow

import * as React from 'react';
import {useCallback} from 'react';
import {View, Pressable, Dimensions} from 'react-native';

import {Colors, INPUT_UPDATE} from '../../../constants';
import {Input, ProfilePicture} from '../../ui';
import styles from './ContactForm.styles';

type State = {
  +type: string,
  +input: string,
  +value: string,
  +isValid: boolean,
  +errorText: string,
};

type Props = {
  formState: {
    inputValues: {
      firstName: string,
      lastName: string,
      age: number,
      photo: String,
    },
  },
  dispatchFormState: (state: State) => void,
  isEditing: boolean,
};

const ContactForm = ({
  formState,
  dispatchFormState,
  isEditing,
}: Props): React.Node => {
  const {inputValues} = formState;
  const {firstName, lastName, age, photo} = inputValues;

  const inputChangeHandler = useCallback(
    (input, value, isValid, errorText) => {
      dispatchFormState({
        type: INPUT_UPDATE,
        input: input,
        value: value,
        isValid: isValid,
        errorText: errorText,
      });
    },
    [dispatchFormState],
  );

  return (
    <View style={styles.form}>
      <Pressable style={styles.profilePictureContainer}>
        <ProfilePicture
          photo={photo}
          firstName={firstName}
          lastName={lastName}
          pictureSize={Dimensions.get('window').height / 8}
          initialsSize={24}
          backgroundColor={Colors.white}
        />
      </Pressable>

      <Input
        id="firstName"
        label="First Name"
        autoCapitalize="words"
        initialValue={firstName}
        initiallyValid={!!isEditing}
        onInputChange={inputChangeHandler}
        required
        alphanumeric
        minLength={3}
        maxLength={30}
      />
      <Input
        id="lastName"
        label="Last Name"
        autoCapitalize="words"
        initialValue={lastName}
        initiallyValid={!!isEditing}
        onInputChange={inputChangeHandler}
        required
        alphanumeric
        minLength={3}
        maxLength={30}
      />

      <View style={styles.row}>
        <Input
          id="age"
          label="Age"
          keyboardType="number-pad"
          onInputChange={inputChangeHandler}
          initialValue={age}
          initiallyValid={!!isEditing}
          required
          numeric
          min={1}
          max={100}
          style={styles.ageInput}
        />
        <Input
          id="photo"
          label="Photo URL"
          keyboardType="url"
          autoCapitalize="none"
          initialValue={photo}
          initiallyValid={true}
          onInputChange={inputChangeHandler}
          style={styles.photoInput}
        />
      </View>
    </View>
  );
};

export default ContactForm;
