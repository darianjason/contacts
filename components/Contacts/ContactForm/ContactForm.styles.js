import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  form: {
    marginBottom: 16,
  },
  profilePictureContainer: {
    marginBottom: 16,
  },
  row: {
    flexDirection: 'row',
    alignItems: 'flex-start',
  },
  ageInput: {
    flex: 1,
    marginRight: 24,
  },
  photoInput: {
    flex: 4,
  },
});

export default styles;
