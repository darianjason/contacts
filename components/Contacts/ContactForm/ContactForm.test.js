import React from 'react';
import renderer from 'react-test-renderer';

import ContactForm from './ContactForm';

describe('Contact Form component', () => {
  test('renders correctly', () => {
    const formState = {
      inputValues: {
        firstName: 'Hello',
        lastName: 'World',
        age: '50',
        photo: 'uri',
      },
    };
    const dispatchFormState = jest.fn();
    const isEditing = jest.fn();

    const tree = renderer
      .create(
        <ContactForm
          formState={formState}
          dispatchFormState={dispatchFormState}
          isEditing={isEditing}
        />,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
