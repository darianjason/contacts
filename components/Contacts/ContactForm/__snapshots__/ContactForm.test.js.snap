// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact Form component renders correctly 1`] = `
<View
  style={
    Object {
      "marginBottom": 16,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginBottom": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#F7F4F3",
          "borderRadius": 8,
          "height": 166.75,
          "justifyContent": "center",
          "overflow": "hidden",
          "width": 166.75,
        }
      }
    >
      <Image
        accessibilityRole="image"
        source={
          Object {
            "uri": "uri",
          }
        }
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
          "marginBottom": 16,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#142B3E",
            "fontFamily": "Manrope-Bold",
            "letterSpacing": 0.5,
            "marginBottom": 8,
          },
          false,
        ]
      }
    >
      First Name
    </Text>
    <TextInput
      alphanumeric={true}
      autoCapitalize="words"
      autoCorrect={false}
      id="firstName"
      initialValue="Hello"
      initiallyValid={true}
      label="First Name"
      maxLength={30}
      minLength={3}
      onBlur={[Function]}
      onChangeText={[Function]}
      onInputChange={[Function]}
      required={true}
      style={
        Array [
          Object {
            "backgroundColor": "#F7F4F3",
            "borderRadius": 4,
            "fontFamily": "Manrope-Regular",
            "paddingHorizontal": 8,
            "paddingVertical": 4,
          },
          false,
        ]
      }
      value="Hello"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
          "marginBottom": 16,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#142B3E",
            "fontFamily": "Manrope-Bold",
            "letterSpacing": 0.5,
            "marginBottom": 8,
          },
          false,
        ]
      }
    >
      Last Name
    </Text>
    <TextInput
      alphanumeric={true}
      autoCapitalize="words"
      autoCorrect={false}
      id="lastName"
      initialValue="World"
      initiallyValid={true}
      label="Last Name"
      maxLength={30}
      minLength={3}
      onBlur={[Function]}
      onChangeText={[Function]}
      onInputChange={[Function]}
      required={true}
      style={
        Array [
          Object {
            "backgroundColor": "#F7F4F3",
            "borderRadius": 4,
            "fontFamily": "Manrope-Regular",
            "paddingHorizontal": 8,
            "paddingVertical": 4,
          },
          false,
        ]
      }
      value="World"
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "justifyContent": "center",
            "marginBottom": 16,
          },
          Object {
            "flex": 1,
            "marginRight": 24,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#142B3E",
              "fontFamily": "Manrope-Bold",
              "letterSpacing": 0.5,
              "marginBottom": 8,
            },
            false,
          ]
        }
      >
        Age
      </Text>
      <TextInput
        id="age"
        initialValue="50"
        initiallyValid={true}
        keyboardType="number-pad"
        label="Age"
        max={100}
        min={1}
        numeric={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onInputChange={[Function]}
        required={true}
        style={
          Array [
            Object {
              "backgroundColor": "#F7F4F3",
              "borderRadius": 4,
              "fontFamily": "Manrope-Regular",
              "paddingHorizontal": 8,
              "paddingVertical": 4,
            },
            false,
          ]
        }
        value="50"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "justifyContent": "center",
            "marginBottom": 16,
          },
          Object {
            "flex": 4,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#142B3E",
              "fontFamily": "Manrope-Bold",
              "letterSpacing": 0.5,
              "marginBottom": 8,
            },
            false,
          ]
        }
      >
        Photo URL
      </Text>
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        id="photo"
        initialValue="uri"
        initiallyValid={true}
        keyboardType="url"
        label="Photo URL"
        onBlur={[Function]}
        onChangeText={[Function]}
        onInputChange={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "#F7F4F3",
              "borderRadius": 4,
              "fontFamily": "Manrope-Regular",
              "paddingHorizontal": 8,
              "paddingVertical": 4,
            },
            false,
          ]
        }
        value="uri"
      />
    </View>
  </View>
</View>
`;
